File: apps/desktop/src/services/commands/importExport.ts
import { invoke } from "@tauri-apps/api/core";

export interface ExportableDeckDto {
  id: number;
  title: string;
  slug: string;
  packName: string;
  wordCount: number;
}

export interface ExportableDecksDto {
  decks: ExportableDeckDto[];
  total: number;
}

export interface ImportExportSchemaDto {
  jsonSchemaName: string;
  jsonSchemaVersion: string;
  jsonRequiredTopLevelFields: string[];
  csvFormatName: string;
  csvHeaders: string[];
  csvNotes: string[];
}

export interface ExportDeckResultDto {
  deckId: number;
  deckSlug: string;
  filePath: string;
  bytesWritten: number;
  wordCount: number;
}

export interface ImportDeckResultDto {
  importId: number;
  packId: number;
  deckId: number;
  deckSlug: string;
  wordsImported: number;
  sensesImported: number;
  examplesImported: number;
  pronunciationsImported: number;
  status: string;
}

export function getImportExportSchema(): Promise<ImportExportSchemaDto> {
  return invoke<ImportExportSchemaDto>("get_import_export_schema");
}

export function listExportableDecks(): Promise<ExportableDecksDto> {
  return invoke<ExportableDecksDto>("list_exportable_decks");
}

export function exportDeckToJson(
  deckId: number,
  filePath: string | null,
  overwrite: boolean,
): Promise<ExportDeckResultDto> {
  return invoke<ExportDeckResultDto>("export_deck_to_json", {
    deckId,
    filePath,
    overwrite,
  });
}

export function importDeckFromJson(
  filePath: string,
): Promise<ImportDeckResultDto> {
  return invoke<ImportDeckResultDto>("import_deck_from_json", { filePath });
}

File: apps/desktop/src-tauri/src/dto/import_export.rs
use serde::Serialize;

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ExportableDeckDto {
    pub id: i64,
    pub title: String,
    pub slug: String,
    pub pack_name: String,
    pub word_count: i64,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ExportableDecksDto {
    pub decks: Vec<ExportableDeckDto>,
    pub total: usize,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ImportExportSchemaDto {
    pub json_schema_name: String,
    pub json_schema_version: String,
    pub json_required_top_level_fields: Vec<String>,
    pub csv_format_name: String,
    pub csv_headers: Vec<String>,
    pub csv_notes: Vec<String>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ExportDeckResultDto {
    pub deck_id: i64,
    pub deck_slug: String,
    pub file_path: String,
    pub bytes_written: u64,
    pub word_count: usize,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ImportDeckResultDto {
    pub import_id: i64,
    pub pack_id: i64,
    pub deck_id: i64,
    pub deck_slug: String,
    pub words_imported: usize,
    pub senses_imported: usize,
    pub examples_imported: usize,
    pub pronunciations_imported: usize,
    pub status: String,
}
